### PR TITLE
FEATURE: add tag-based gated content

### PR DESCRIPTION
**Pull request by chapoi**

### CODE REVIEW

File: settings.yml [hunk 1]
```diff
@@ -3,3 +3,8 @@ enabled_categories:
   list_type: category
   default: ""
   description: "Choose which categories that users need to sign up for."
+enabled_tags:
+  type: list
+  list_type: tag
+  default: "gated"
```
Comment (davidtaylorhq) [inline settings.yml:9]: Let's make the default empty, so it definitely doesn't affect existing sites

```suggestion
  default: ""
```

File: javascripts/discourse/components/topic-in-gated-category.js [hunk 2]
```diff
@@ -26,18 +29,20 @@ export default Component.extend({
 
   recalculate() {
     // do nothing if:
-    // a) topic does not have a category
+    // a) topic does not have a category and does not have a gated tag
     // b) component setting is empty
     // c) user is logged in
+    this.gatedByTag = this.tags.some((t) => enabledTags.indexOf(t) >= 0);
```
Comment (davidtaylorhq) [inline javascripts/discourse/components/topic-in-gated-category.js:35]: Functionality is the same, but I think I prefer the readability of
```suggestion
    this.gatedByTag = this.tags.some((t) => enabledTags.includes(t));
```

File: javascripts/discourse/components/topic-in-gated-category.js [hunk 3]
```diff
@@ -6,9 +6,12 @@ const enabledCategories = settings.enabled_categories
   .map((id) => parseInt(id, 10))
   .filter((id) => id);
 
+const enabledTags = settings.enabled_tags.split("|");
+
 export default Component.extend({
   tagName: "",
   hidden: true,
+  gatedByTag: false,
```
Comment (davidtaylorhq) [inline javascripts/discourse/components/topic-in-gated-category.js:14]: Is there any need for this to be a property of the component? Looks like it's only used inside the `recalculate()` method, so maybe we should just make it a local variable in there?

File: javascripts/discourse/components/topic-in-gated-category.js [hunk 4]
```diff
@@ -6,9 +6,12 @@ const enabledCategories = settings.enabled_categories
   .map((id) => parseInt(id, 10))
   .filter((id) => id);
 
+const enabledTags = settings.enabled_tags.split("|");
```
Comment (davidtaylorhq) [inline javascripts/discourse/components/topic-in-gated-category.js:9]: If the setting is an empty string, and we call `.split("|")`, it'll end up as an array like

```
[""]
```
which is kinda weird (and breaks the `.length === 0` check later). We can filter out empty values using the same technique as the `enabledCategory` variable above

```suggestion
const enabledTags = settings.enabled_tags.split("|").filter((t) => t);
```

**Review by davidtaylorhq — COMMENTED** [review]
Looks great to me! Just a few style comments

I guess we should think about renaming this component to 'discourse-gated-topics' 😅

**Comment by chapoi** [comment]
> Looks great to me! Just a few style comments
> 
> I guess we should think about renaming this component to 'discourse-gated-topics' 😅

Yeah I agree, but can we do that without interfering with current usage?

**Comment by jbrw** [comment]
Note that the second commit assumes that https://github.com/discourse/discourse/pull/18162 has been merged.

**Comment by pmusaraj** [comment]
> Yeah I agree, but can we do that without interfering with current usage?

I think so. Github redirects moved repos, so we should move to `https://github.com/discourse/discourse-gated-topics`

File: javascripts/discourse/components/topic-in-gated-category.js [hunk 5]
```diff
@@ -6,6 +6,8 @@ const enabledCategories = settings.enabled_categories
   .map((id) => parseInt(id, 10))
   .filter((id) => id);
 
+const enabledTags = settings.enabled_tags.split("|").filter((t) => t);
```
Comment (jjaffeux) [inline javascripts/discourse/components/topic-in-gated-category.js:9]: Idiomatic way of writing this is:

```
const enabledTags = settings.enabled_tags.split("|").filter(Boolean);
```

But up to you, totally fine like this.

**Review by jjaffeux — APPROVED** [review]

**Review by pmusaraj — APPROVED** [review]

**Pull request merged**